ObserverSet.swift

```diff
@@ -26,122 +26,198 @@
+	- parameter keyPath: The keyPath to set when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which can be used later to remove the observer.
+	*/
+	@discardableResult
+	open func add<ObserverType: AnyObject>(queue: DispatchQueue? = nil, _ observer: ObserverType, _ keyPath: ReferenceWritableKeyPath<ObserverType, Parameters>) -> ObserverSetEntry<Parameters> {
+		add(queue: queue, observer) { observer in { observer[keyPath: keyPath] = $0 } }
+	}
+	
+	/**
+	Removes an observer from the list, using the entry which was returned when adding.
+	
+	- parameter entry: An entry returned when adding a new observer.
+	*/
+	open func remove(_ entry: ObserverSetEntry<Parameters>) {
+		synchronized {
+			self.entries = self.entries.filter{ $0 !== entry }
+		}
+	}
+	
+	
+	/**
+	Removes an observer from the list.
+	
+	- parameter observer: An observer to remove from the list of observers.
+	*/
+	open func removeObserver<ObserverType: AnyObject>(_ observer: ObserverType) {
+		synchronized {
+			self.entries.removeAll(where: { ($0 as? ObserverSetEntryImpl<ObserverType, Parameters>)?.observer === observer })
+		}
+	}
+	
+	/**
+	Call this method to notify all observers.
+	
+	- parameter parameters: The parameters that are required parameters specified using generics when the instance is created.
+	*/
+	open func notify(_ parameters: Parameters) {
+		let callBackList = synchronized { () -> [ObserverSetEntry<Parameters>] in
+			self.entries.removeAll(where: { $0.isExpired })
+			return self.entries
+		}
+		for callBack in callBackList {
+			callBack.call(parameters)
+		}
+	}
+	
+	// MARK: - Private functions
+	// MARK: Locking support
+	
+	private var queue = DispatchQueue(label: "com.theappbusiness.ObserverSet", attributes: [])
+	
+	private func synchronized<T>(_ f: () -> T) -> T {
+		queue.sync(execute: f)
+	}
 }
 
 public extension ObserverSet where Parameters == Void {
-  public func notify() {
-    notify(())
-  }
+	
+	/**
+	Call this method to notify all observers.
+	*/
+	func notify() {
+		notify(())
+	}
+	/**
+	Adds a new observer to the set.
+	
+	Convenience method so that the callback function can be `func callback() { ... }` instead of `func callback(_: Void) { ... }`
+
+	- parameter queue: The queue to call the function on when the notification is delivered. If nil (the default) it will be delivered on the same queue the notification was sent.
+	- parameter observer: The object that is to be notified.
+	- parameter callBack: The function to call on the observer when the notification is to be delivered.
+	
+	- returns: An entry in the list of observers, which can be used later to remove the observer.
+	*/
+	@discardableResult
+	func add<ObserverType: AnyObject>(queue: DispatchQueue? = nil, _ observer: ObserverType, _ callBack: @escaping (ObserverType) -> () -> Void) -> ObserverSetEntry<Parameters> {
```

Comment:
This isn't needed for the other adds as Swift seems to realise `() -> Void` and `(Void) -> Void` are the same thing